### PR TITLE
Gardenlet creates the Seed as early as possible

### LINKED ISSUE #3816: Gardenlet panics when deployed to a managed seed
**How to categorize this issue?**

/area control-plane
/kind bug
/priority 3

**What happened**:

When deployed to a managed seed via a `ManagedSeed` resource, gardenlet panics:

```
panic: need an internal domain secret but found none

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.Must(...)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:171
github.com/gardener/gardener/pkg/gardenlet/controller.(*GardenletControllerFactory).Run(0xc000f079b8, 0x256c860, 0xc0006f1280, 0xc000254300, 0xc00005e800)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/factory.go:145 +0x1ce5
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).startControllers(0xc000991720, 0x256c860, 0xc0006f1280, 0x252d101, 0x251a740)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:493 +0x152
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).Run.func1(0x256c860, 0xc0006f1280, 0x256c860, 0xc0006f07c0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:445 +0x65
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).Run(0xc000991720, 0x256c860, 0xc00052ecc0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:475 +0x790
github.com/gardener/gardener/cmd/gardenlet/app.run(0x256c860, 0xc00052ecc0, 0xc0003c08a0, 0x1db7840, 0xc0007d5d60)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:171 +0x10a
github.com/gardener/gardener/cmd/gardenlet/app.NewCommandStartGardenlet.func1(0xc0007ec840, 0xc00028e6a0, 0x0, 0x1, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:201 +0x99
github.com/spf13/cobra.(*Command).execute(0xc0007ec840, 0xc00004c070, 0x1, 0x1, 0xc0007ec840, 0xc00004c070)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:850 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007ec840, 0xc00052ecc0, 0x0, 0x1)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:895
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:888
main.main()
	/go/src/github.com/gardener/gardener/cmd/gardenlet/main.go:36 +0x12c
```

**What you expected to happen**:
No panic.

**How to reproduce it (as minimally and precisely as possible)**:
Create a `ManagedSeed` resource with `gardenlet`, or use the `use-as-seed` annotation.

**Anything else we need to know?**:

**Environment**:

- Gardener version: master (d6914db40e716cb1e60c5e726f64ffea4405d8ee)
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by vpnachev**
This is required so that the GCM can see the new Seed to
- create the seed namespace
- replicate the resources from the garden to the seed namespace

**How to categorize this PR?**

/kind bug

**What this PR does / why we need it**:
Gardenlet creates the Seed as early as possible.

**Which issue(s) this PR fixes**:
Fixes #3816

**Special notes for your reviewer**:

**Release note**:

```bugfix operator
The Gardenlet is now creating/updating the Seed resource from `GardenletConfiguration.seedConfig` earlier in the start-up flow to allow Gardener Controller Manager to replicate the required credentials in the namespace dedicated to the configured seed.
```

/invite @stoyanr

### CODE REVIEW

File: pkg/gardenlet/controller/factory.go [hunk 1]
```diff
@@ -134,8 +136,30 @@ func (f *GardenletControllerFactory) Run(ctx context.Context) error {
 		return fmt.Errorf("no seed selected by Gardenlet")
 	}
 
+	// Register Seed object if desired
+	if f.cfg.SeedConfig != nil {
+		// Convert gardenlet config to an external version
+		cfg, err := confighelper.ConvertGardenletConfigurationExternal(f.cfg)
+		if err != nil {
+			panic(fmt.Errorf("could not convert gardenlet configuration: %+v", err))
```
Comment (ialidzhikov) [inline pkg/gardenlet/controller/factory.go]:
```suggestion
			return fmt.Errorf("could not convert gardenlet configuration: %+v", err)
```

File: pkg/gardenlet/controller/factory.go [hunk 2]
```diff
@@ -134,8 +136,30 @@ func (f *GardenletControllerFactory) Run(ctx context.Context) error {
...
+		seed := &gardencorev1beta1.Seed{ObjectMeta: metav1.ObjectMeta{Name: f.cfg.SeedConfig.Name}}
+		if _, err := controllerutil.CreateOrUpdate(ctx, k8sGardenClient.Client(), seed, func() error {
+			seed.Labels = utils.MergeStringMaps(map[string]string{
+				v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
+			}, f.cfg.SeedConfig.Labels)
+
+			seed.Spec = cfg.SeedConfig.Spec
+			return nil
+		}); err != nil {
+			panic(fmt.Errorf("could not register seed %q: %+v", seed.Name, err))
```
Comment (ialidzhikov) [inline pkg/gardenlet/controller/factory.go]:
```suggestion
			return fmt.Errorf("could not register seed %q: %+v", seed.Name, err)
```

**Comment by stoyanr** [comment]
I can confirm that with this change, the error reported in #3816 is no longer reproducible for a seed created by a `gardenlet` managed seed.

**Review by stoyanr — COMMENTED** [review]
/lgtm except for the changes already requested by @ialidzhikov

File: pkg/gardenlet/controller/factory.go [hunk 3]
```diff
@@ -134,8 +136,30 @@ func (f *GardenletControllerFactory) Run(ctx context.Context) error {
...
+		}
+	}
+
 	// Read Garden secrets from any seed namespace as we assume they are synced accordingly by the Gardener Controller Manager.
 	// This requires adaption if we'll decide not to sync all Secrets for all Seeds in the future.
+	// For newly created seed, this seed namespace is about to be created by the GCM.
 	_, err = garden.ReadGardenSecrets(
```
Comment (ialidzhikov) [inline pkg/gardenlet/controller/factory.go:153]: As these secrets are created by external component (gardener-controller-manager), we should wait for some short period of time until the secrets are created in the seed namespace. Currently without such wait the Seed registration is likely to fail in the first 1-2 attempts and succeed after these attempts.

Comment (vpnachev) [reply pkg/gardenlet/controller/factory.go:153]: I have added such wait mechanism for the namespace.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by stoyanr — APPROVED** [review]

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**